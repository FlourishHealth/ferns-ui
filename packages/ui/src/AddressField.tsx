import React, {FC} from "react";
import {View} from "react-native";

import {AddressFieldProps, AddressInterface} from "./Common";
import {USSTATESLIST} from "./Constants";
import {SelectField} from "./SelectField";
import {TextField} from "./TextField";
import {UnifiedAddressAutoCompleteField} from "./UnifiedAddressAutoComplete";

export const AddressField: FC<AddressFieldProps> = ({
  disabled,
  googleMapsApiKey,
  googlePlacesMobileStyles,
  includeCounty,
  value,
  testID,
  onChange,
  onBlur,
}) => {
  const handleAddressChange = (field: string, newValue: string) => {
    onChange({...value, [field]: newValue});
    onBlur && onBlur({...value, [field]: newValue});
  };

  const handleAutoCompleteChange = (newValue: AddressInterface) => {
    onChange({...value, ...newValue});
  };

  const {
    address1 = "",
    address2 = "",
    city = "",
    state = "",
    zipcode = "",
    countyName = "",
    countyCode = "",
  }: AddressInterface = value ?? ({} as AddressInterface);

  return (
    <>
      <View style={{marginBottom: 16}}>
        <UnifiedAddressAutoCompleteField
          disabled={disabled}
          googleMapsApiKey={googleMapsApiKey}
          googlePlacesMobileStyles={googlePlacesMobileStyles}
          handleAddressChange={(result) => handleAddressChange("address1", result)}
          handleAutoCompleteChange={(result) => handleAutoCompleteChange(result)}
          includeCounty={includeCounty}
          inputValue={address1}
          testID={`${testID}-address1`}
        />
      </View>
      <View style={{marginBottom: 16}}>
        <TextField
          disabled={disabled}
          id="address2"
          testID={`${testID}-address2`}
          title="Apt, suite, etc"
          type="text"
          value={address2}
          onChange={(result) => handleAddressChange("address2", result)}
        />
      </View>
      <View style={{marginBottom: 16}}>
        <TextField
          disabled={disabled}
          id="city"
          testID={`${testID}-city`}
          title="City"
          type="text"
          value={city}
          onChange={(result) => handleAddressChange("city", result)}
        />
      </View>
      <View style={{marginBottom: 16}}>
        <SelectField
          options={USSTATESLIST}
          title="State"
          value={state}
          onChange={(result) => handleAddressChange("state", result!)}
        />
      </View>
      <View style={{marginBottom: 16}}>
        <TextField
          disabled={disabled}
          id="zipcode"
          testID={`${testID}-zip`}
          title="Zipcode"
          type="text"
          value={zipcode}
          onChange={(result) => handleAddressChange("zipcode", result)}
        />
      </View>
      {includeCounty && (
        <>
          <View style={{marginBottom: 16}}>
            <TextField
              disabled={disabled}
              id="countyName"
              testID={`${testID}-county`}
              title="County Name"
              type="text"
              value={countyName}
              onChange={(result) => handleAddressChange("countyName", result)}
            />
          </View>
          <View style={{marginBottom: 16}}>
            <TextField
              disabled={disabled}
              id="countyCode"
              testID={`${testID}-county-code`}
              title="County Code"
              type="text"
              value={countyCode}
              onChange={(result) => handleAddressChange("countyCode", result)}
            />
          </View>
        </>
      )}
    </>
  );
};
