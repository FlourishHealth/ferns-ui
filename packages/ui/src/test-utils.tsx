import {render} from "@testing-library/react-native";
import React from "react";
import {ThemeProvider} from "./Theme";

export const renderWithTheme = (ui: React.ReactElement) => {
  return render(<ThemeProvider>{ui}</ThemeProvider>);
};

export const createCommonMocks = () => ({
  onChange: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  onEnter: jest.fn(),
  onSubmitEditing: jest.fn(),
  onIconClick: jest.fn(),
});

export const setupComponentTest = () => {
  jest.useFakeTimers();
  return createCommonMocks();
};

export const teardownComponentTest = () => {
  jest.useRealTimers();
  jest.clearAllMocks();
};
