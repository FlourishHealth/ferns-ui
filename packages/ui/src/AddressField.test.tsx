import {fireEvent} from "@testing-library/react-native";
import React from "react";

import {AddressField} from "./AddressField";
import {renderWithTheme} from "./test-utils";

describe("AddressField", () => {
  const defaultProps = {
    testID: "test-address",
    value: {
      address1: "123 Main St",
      address2: "Apt 4B",
      city: "Springfield",
      state: "IL",
      zipcode: "62701",
      countyName: "Sangamon",
      countyCode: "17167",
    },
    onChange: jest.fn(),
    onBlur: jest.fn(),
  };

  it("renders correctly with default props", () => {
    const {toJSON} = renderWithTheme(<AddressField {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders all address fields", () => {
    const {getByTestId} = renderWithTheme(<AddressField {...defaultProps} />);

    expect(getByTestId("test-address-address1")).toBeTruthy();
    expect(getByTestId("test-address-address2")).toBeTruthy();
    expect(getByTestId("test-address-city")).toBeTruthy();
    expect(getByTestId("test-address-zip")).toBeTruthy();
  });

  it("calls onChange when address fields are updated", () => {
    const {getByTestId} = renderWithTheme(<AddressField {...defaultProps} />);

    const cityInput = getByTestId("test-address-city");
    fireEvent.changeText(cityInput, "New City");

    expect(defaultProps.onChange).toHaveBeenCalledWith({
      ...defaultProps.value,
      city: "New City",
    });
  });

  it("calls onBlur when a field is blurred", () => {
    const {getByTestId} = renderWithTheme(<AddressField {...defaultProps} />);
    // Reset mock since it was called during render
    defaultProps.onBlur.mockClear();
    const zipInput = getByTestId("test-address-zip");
    fireEvent.changeText(zipInput, "90210");
    fireEvent(zipInput, "blur");

    expect(defaultProps.onBlur).toHaveBeenCalledWith({
      ...defaultProps.value,
      zipcode: "90210",
    });
  });

  it("renders county fields when includeCounty is true", () => {
    const {getByTestId} = renderWithTheme(<AddressField {...defaultProps} includeCounty />);

    expect(getByTestId("test-address-county")).toBeTruthy();
    expect(getByTestId("test-address-county-code")).toBeTruthy();
  });

  it("does not render county fields when includeCounty is false", () => {
    const {queryByTestId} = renderWithTheme(
      <AddressField {...defaultProps} includeCounty={false} />
    );

    expect(queryByTestId("test-address-county")).toBeNull();
    expect(queryByTestId("test-address-county-code")).toBeNull();
  });

  it("disables all fields when disabled prop is true", () => {
    const {getByTestId} = renderWithTheme(<AddressField {...defaultProps} disabled />);

    const address1Input = getByTestId("test-address-address1");
    const cityInput = getByTestId("test-address-city");

    // Check that the disabled prop is passed down to the inputs
    expect(address1Input.props.accessibilityState.disabled).toBe(true);
    expect(cityInput.props.accessibilityState.disabled).toBe(true);
  });
});
