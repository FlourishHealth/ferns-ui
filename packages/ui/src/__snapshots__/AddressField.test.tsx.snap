// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressField renders correctly with default props 1`] = `
[
  <View
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#1C1C1C",
            "fontSize": 14,
            "fontWeight": 600,
            "lineHeight": 22.4,
          }
        }
      >
        Street Address
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#9A9A9A",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "overflow": "hidden",
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          }
        }
      >
        <TextInput
          accessibilityHint="Enter text here"
          accessibilityState={
            {
              "disabled": undefined,
            }
          }
          aria-label="Text input field"
          autoCapitalize="sentences"
          autoCorrect={true}
          blurOnSubmit={true}
          keyboardType="default"
          numberOfLines={1}
          onBlur={[Function]}
          onChangeText={[Function]}
          onContentSizeChange={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Enter an address"
          placeholderTextColor="#686868"
          secureTextEntry={false}
          style={
            {
              "color": "#1C1C1C",
              "flex": 1,
              "fontFamily": "text",
              "fontSize": 16,
              "gap": 10,
              "height": 20,
              "paddingVertical": 0,
              "width": "100%",
            }
          }
          testID="test-address-address1"
          textContentType="none"
          underlineColorAndroid="transparent"
          value="123 Main St"
        />
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#1C1C1C",
            "fontSize": 14,
            "fontWeight": 600,
            "lineHeight": 22.4,
          }
        }
      >
        Apt, suite, etc
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#9A9A9A",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "overflow": "hidden",
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          }
        }
      >
        <TextInput
          accessibilityHint="Enter text here"
          accessibilityState={
            {
              "disabled": undefined,
            }
          }
          aria-label="Text input field"
          autoCapitalize="sentences"
          autoCorrect={true}
          blurOnSubmit={true}
          keyboardType="default"
          numberOfLines={1}
          onBlur={[Function]}
          onChangeText={[Function]}
          onContentSizeChange={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholderTextColor="#686868"
          secureTextEntry={false}
          style={
            {
              "color": "#1C1C1C",
              "flex": 1,
              "fontFamily": "text",
              "fontSize": 16,
              "gap": 10,
              "height": 20,
              "paddingVertical": 0,
              "width": "100%",
            }
          }
          testID="test-address-address2"
          textContentType="none"
          underlineColorAndroid="transparent"
          value="Apt 4B"
        />
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#1C1C1C",
            "fontSize": 14,
            "fontWeight": 600,
            "lineHeight": 22.4,
          }
        }
      >
        City
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#9A9A9A",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "overflow": "hidden",
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          }
        }
      >
        <TextInput
          accessibilityHint="Enter text here"
          accessibilityState={
            {
              "disabled": undefined,
            }
          }
          aria-label="Text input field"
          autoCapitalize="sentences"
          autoCorrect={true}
          blurOnSubmit={true}
          keyboardType="default"
          numberOfLines={1}
          onBlur={[Function]}
          onChangeText={[Function]}
          onContentSizeChange={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholderTextColor="#686868"
          secureTextEntry={false}
          style={
            {
              "color": "#1C1C1C",
              "flex": 1,
              "fontFamily": "text",
              "fontSize": 16,
              "gap": 10,
              "height": 20,
              "paddingVertical": 0,
              "width": "100%",
            }
          }
          testID="test-address-city"
          textContentType="none"
          underlineColorAndroid="transparent"
          value="Springfield"
        />
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <View>
      <Text
        style={
          {
            "color": "#1C1C1C",
            "fontSize": 14,
            "fontWeight": 600,
            "lineHeight": 22.4,
          }
        }
      >
        State
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "alignSelf": "stretch",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 40,
              "width": "100%",
            },
            {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#9A9A9A",
            },
            false,
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          activeOpacity={1}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 40,
              "width": "95%",
            }
          }
          testID="ios_touchable_wrapper"
        >
          <View
            style={
              {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "pointerEvents": "box-only",
                "width": "100%",
              }
            }
          >
            <TextInput
              readOnly={true}
              style={
                {
                  "color": "#1C1C1C",
                }
              }
              testID="text_input"
              value="Please select an option."
            />
            <View
              style={
                {
                  "pointerEvents": "none",
                }
              }
              testID="icon_container"
            >
              <View
                brand={false}
                color="#1C1C1C"
                duotone={false}
                light={false}
                name="angle-down"
                regular={false}
                sharp={false}
                size={12}
                solid={true}
                thin={false}
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "column",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#1C1C1C",
            "fontSize": 14,
            "fontWeight": 600,
            "lineHeight": 22.4,
          }
        }
      >
        Zipcode
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#9A9A9A",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "overflow": "hidden",
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          }
        }
      >
        <TextInput
          accessibilityHint="Enter text here"
          accessibilityState={
            {
              "disabled": undefined,
            }
          }
          aria-label="Text input field"
          autoCapitalize="sentences"
          autoCorrect={true}
          blurOnSubmit={true}
          keyboardType="default"
          numberOfLines={1}
          onBlur={[Function]}
          onChangeText={[Function]}
          onContentSizeChange={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholderTextColor="#686868"
          secureTextEntry={false}
          style={
            {
              "color": "#1C1C1C",
              "flex": 1,
              "fontFamily": "text",
              "fontSize": 16,
              "gap": 10,
              "height": 20,
              "paddingVertical": 0,
              "width": "100%",
            }
          }
          testID="test-address-zip"
          textContentType="none"
          underlineColorAndroid="transparent"
          value="62701"
        />
      </View>
    </View>
  </View>,
]
`;
